### PR TITLE
Flesh out a basic README for Alloy

### PR DESCRIPTION
**Pull request by jacob-hughes**
This is just a start, it contains the basic build steps and a (hopefully) motivating example. This still needs to explain the intricacies of finalisers.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -111,170 +106,53 @@ See [the rustc-dev-guide for more info][sysllvm].
...
+Alloy uses _conservative_ garbage collection. This means that it does not have
+any specific knowledge about where references to objects are located. Instead,
+Alloy will assume that an object is still alive if it can be reached by a value on
+the stack (or in a register) which, if treated like a pointer, points to an
+object in the heap. The fields of those objects are then traced using the same
+approach, until all live objects in the program have been discovered. 
 
-## Building Documentation
+This tends to work well in practice, however, it comes with an important caveat:
+you must not hide references from the type system. For example, data structures
```
Comment (ltratt) [inline README.md:134]: s/from the type system/from the GC/

File: README.md [hunk 2]
```diff
@@ -111,170 +106,53 @@ See [the rustc-dev-guide for more info][sysllvm].
...
+such as XOR lists are unsound because Alloy will never be able to reach their
+objects.
 
-If you'd like to build the documentation, it's almost the same:
+Behind the scenes, Alloy uses the [Boehm Demers Weiser GC (BDWGC)] for its
```
Comment (ltratt) [inline README.md:138]: Missing URL in brackets for BDWGC?

Comment (jacob-hughes) [reply README.md:138]: No this is just weird Markdown syntax. The link is on line +124

File: README.md (hunk 1 above)
Comment (jacob-hughes) [reply README.md]: Fixed here 
1eb9bba

**Comment by jacob-hughes** [comment]
Can I squash this?

File: README.md (hunk 2 above)
Comment (ltratt) [reply README.md:138]: Oh!

**Comment by ltratt** [comment]
Please squash.

**Comment by jacob-hughes** [comment]
Squashed

**Pull request merged**